File: mcp-server/src/preview/rpcAllowlist.ts
import { loadPreviewConfig } from "./previewConfig.js";
import { discoverWorkspaceRpcMethods } from "./rpcDiscover.js";

export type RpcAllowlistSnapshot = {
  rpcMode: string;
  allowed: string[];
  denylist: string[];
  discovered: string[];
  generatedAt: number;
};

export function resolveRpcAllowlist(workspaceRoot?: string): RpcAllowlistSnapshot {
  const config = loadPreviewConfig(workspaceRoot);
  const discovered = discoverWorkspaceRpcMethods(workspaceRoot);
  const deny = new Set(config.rpcDenylist);

  if (config.rpcMode === "dev") {
    console.warn(
      "[decky-plugin-studio] preview RPC mode is 'dev' — all methods allowed except denylist"
    );
    return {
      rpcMode: "dev",
      allowed: ["*"],
      denylist: [...deny],
      discovered,
      generatedAt: Date.now(),
    };
  }

  if (config.rpcMode === "allowlist") {
    const allowed = config.rpcAllowlist.filter((m) => !deny.has(m) && !m.startsWith("_"));
    return {
      rpcMode: "allowlist",
      allowed,
      denylist: [...deny],
      discovered,
      generatedAt: Date.now(),
    };
  }

  // discover (default): discovered + explicit allowlist extras
  const merged = new Set<string>();
  for (const m of discovered) {
    if (!deny.has(m) && !m.startsWith("_")) merged.add(m);
  }
  for (const m of config.rpcAllowlist) {
    if (!deny.has(m) && !m.startsWith("_")) merged.add(m);
  }

  return {
    rpcMode: "discover",
    allowed: [...merged].sort(),
    denylist: [...deny],
    discovered,
    generatedAt: Date.now(),
  };
}

export function isPreviewRpcAllowed(method: string, workspaceRoot?: string): boolean {
  const snap = resolveRpcAllowlist(workspaceRoot);
  if (snap.denylist.includes(method) || method.startsWith("_")) return false;
  if (snap.rpcMode === "dev") return true;
  if (snap.allowed.includes("*")) return true;
  return snap.allowed.includes(method);
}

/** @deprecated use isPreviewRpcAllowed */
export function isPreviewRpcAllowedLegacy(method: string): boolean {
  return isPreviewRpcAllowed(method);
}

File: extension/resources/preview-server/python/decky_shim/rpc_allowlist.py
"""Load preview RPC allowlist from sandbox preview-rpc.json (written by MCP / extension)."""
from __future__ import annotations

import json
import logging
from pathlib import Path

logger = logging.getLogger("decky-sidecar.rpc_allowlist")

DEFAULT_DENYLIST = frozenset({"_main", "_unload", "_migration"})


def _sandbox_rpc_path() -> Path:
    import os

    sandbox = Path(os.environ.get("DECKY_SANDBOX_ROOT", Path.home() / ".decky-plugin-studio" / "sandbox" / "plugin"))
    return sandbox / "preview-rpc.json"


def load_rpc_policy() -> tuple[str, frozenset[str], frozenset[str]]:
    """
    Returns (rpc_mode, allowed_set, denylist).
    allowed_set contains '*' when dev mode allows all non-denied methods.
    """
    path = _sandbox_rpc_path()
    if not path.exists():
        logger.warning("No preview-rpc.json at %s — denying all RPC until synced", path)
        return "missing", frozenset(), DEFAULT_DENYLIST

    try:
        data = json.loads(path.read_text(encoding="utf-8"))
    except Exception as exc:
        logger.error("Failed to read preview-rpc.json: %s", exc)
        return "error", frozenset(), DEFAULT_DENYLIST

    mode = str(data.get("rpcMode", "discover"))
    deny = frozenset(data.get("denylist", [])) | DEFAULT_DENYLIST
    allowed_raw = data.get("allowed", [])
    allowed = frozenset(allowed_raw) if isinstance(allowed_raw, list) else frozenset()

    if mode == "dev":
        return "dev", frozenset({"*"}), deny

    return mode, allowed, deny


def is_rpc_allowed(method: str) -> bool:
    if not method or method.startswith("_"):
        return False
    mode, allowed, deny = load_rpc_policy()
    if method in deny:
        return False
    if mode == "dev" or "*" in allowed:
        return True
    if mode == "missing":
        return False
    return method in allowed
